core/bin/zksync_tee_prover/src/main.rs

```diff
@@ -0,0 +1,356 @@
+        "tee_prover_layer"
+    }
+
+    async fn wire(self: Box<Self>, mut context: ServiceContext<'_>) -> Result<(), WiringError> {
+        let tee_prover_task = TeeProverTask {
+            api_url: self.api_url,
+            signing_key: self.signing_key,
+            attestation_quote_bytes: self.attestation_quote_bytes,
+            batch_count: self.batch_count,
+            http_client: Client::new(),
+        };
+        context.add_task(Box::new(tee_prover_task));
+        Ok(())
+    }
+}
+
+#[derive(Debug, Parser)]
+#[command(author = "Matter Labs", version, about = "TEE prover", long_about = None)]
+struct Cli {
+    // The URL of the TEE prover interface API.
+    #[arg(long)]
+    api_url: String,
+
+    /// Specifies the number of batches to process. If not provided, the process will continue until
+    /// it is interrupted by a SIGINT signal (CTRL+C).
+    #[arg(long)]
+    batch_count: Option<usize>,
+}
+
+/// This application is a TEE verifier (a.k.a. a prover, or worker) that interacts with three
+/// endpoints of the TEE prover interface API:
+/// 1. `/tee/proof_inputs` - Fetches input data about a batch for the TEE verifier to process.
+/// 2. `/tee/submit_proofs/<l1_batch_number>` - Submits the TEE proof, which is a signature of the
+///    root hash.
+/// 3. `/tee/register_attestation` - Registers the TEE attestation that binds a key used to sign a
+///    root hash to the enclave where the signing process occurred. This effectively proves that the
+///    signature was produced in a trusted execution environment.
+///
+/// Conceptually it works as follows:
+/// 1. Get the TEE_SIGNING_KEY private key from the environment variable.
+/// 2. Get the file path for the file containing the TEE quote from the TEE_QUOTE_FILE environment
+///    variable.
+/// 3. Register the attestation via the `/tee/register_attestation` endpoint.
+/// 4. Run a loop:
+///    a. Fetch the next batch data via the `/tee/proof_inputs` endpoint.
+///    b. Verify the batch data.
+///    c. If verification is successful, sign the root hash of the batch data with the private key.
+///    d. Submit the signature (a.k.a. proof) via the `/tee/submit_proofs/<l1_batch_number>`
+///       endpoint.
+fn main() -> anyhow::Result<()> {
+    let observability_config =
+        ObservabilityConfig::from_env().context("ObservabilityConfig::from_env()")?;
+    let log_format: zksync_vlog::LogFormat = observability_config
+        .log_format
+        .parse()
+        .context("Invalid log format")?;
+    let mut builder = zksync_vlog::ObservabilityBuilder::new().with_log_format(log_format);
+    if let Some(sentry_url) = observability_config.sentry_url {
+        builder = builder
+            .with_sentry_url(&sentry_url)
+            .context("Invalid Sentry URL")?
+            .with_sentry_environment(observability_config.sentry_environment);
+    }
+    let _guard = builder.build();
+
+    let signing_key_pem = std::env::var("TEE_SIGNING_KEY").unwrap_or(
```

Comment:
remove the unwrap_or... this was just for testing

Reply:
It's still for me to test. :P I will certainly remove it! 👍🏼

Reply:
[Done](https://github.com/matter-labs/zksync-era/pull/2333/commits/c7772e8ae58ecfd62eb9b15a9c496edfec935199#diff-30ff3f1eaef9e2673b77750e19ea5af9672ffc5e81c59e8740a1b371098b6183L300-L309). ✅